enhancements/sdk-generate-commands.md

```diff
@@ -0,0 +1,411 @@
+
+New project layout defaults:
+* Output directory: `config/bundle/metadata`
+* Read from: _none_
+
+Old project layout defaults:
+* Output directory: `deploy/olm-catalog/<operator-name>/metadata`
+* Read from: _none_
+
+Notable flags: `--channels=[channel-names]`, `--default-channel=<channel-name>`, `--overwrite=<bool>`
+
+#### `genenerate packagemanifests`
+
+All files/manifests will be generated under `config/packages`, ex. `config/packages/0.0.1`.
+The functionality of this command is split up into steps by flag:
+`--kustomize` (for new project layouts only) and `--manifests`.
+
+###### `--kustomize` (new project layouts)
+
+Generate a `kustomize` base CSV for a new project layout under `config/packages/bases`.
+This base will be used as an actual base for `kustomize build` eventually; for now
+only `generate packagemanifests` will read this base. This step involves generating UI
+metadata from API types, although `--manifests` can do this as well. This subcommand
+is interactive if a base does not exist.
+
+New project layout defaults:
+* Output directory: `config/packages/bases`
+* Read from: `--apis-dir="api(s)"`
+
+###### `--manifests`
+
+Generate a package containing a CSV and owned CRD manifests, and updates the
+[package manifest file][package-manifest-doc]. This step uses a base generated
+with `--kustomize`, or creates a new one. This subcommand is interactive if a
+base does not exist.
+
+New project layout defaults:
+* Output directory: `config/packages/<version>`
+* Read from: stdin, `--apis-dir="api(s)"`
+
+Old project layout defaults:
+* Output directory: `deploy/olm-catalog/<operator-name>/<version>`
+* Read from: `--manifest-root="deploy"`, `--crds-dir="deploy/crds"`, `--apis-dir="pkg/apis"`
+
+Notable flags: `--version=<semver>`, `--channel=<channel-name>`, `--default-channel=<channel-name>`
+
+#### `make` recipes (new project layouts)
+
+New project layouts are scaffolded with a [Makefile][kb-makefile] containing a
+set of pre-defined recipes for generating manifests, building images, etc. To align
+with this idiom, recipes should be amended/added to include `generate bundle` by
+default. If a user does not want these features turned on in their project, they
+can manually remove the amendments/additions.
+
+###### Modifying the base scaffold
+
+`operator-sdk init` will scaffold a Makefile containing the following amendments/additions:
+
+```make
+# Current Operator version
+VERSION ?= 0.0.1
+# Default bundle image tag
+BUNDLE_IMG ?= controller-bundle:$(VERSION)
+# Options for 'bundle-build'
+ifeq ($(origin BUNDLE_CHANNELS), undefined)
+BUNDLE_CHANNELS = --channels=$(BUNDLE_CHANNELS)
+endif
+ifeq ($(origin BUNDLE_DEFAULT_CHANNEL), undefined)
+BUNDLE_DEFAULT_CHANNEL := --default-channel=$(BUNDLE_DEFAULT_CHANNEL)
+endif
+BUNDLE_METADATA_OPTS ?= $(BUNDLE_CHANNELS) $(BUNDLE_DEFAULT_CHANNEL)
+
+...
+
+# Existing manifests recipe, append 'generate bundle'
+manifests:
+  ...
+  operator-sdk generate bundle --kustomize
```

Comment:
Summarizing the UX discussion over from https://github.com/operator-framework/operator-sdk/pull/2860
/cc @camilamacedo86 

If I'm trying to write an operator without intending to write CSVs or bundles, with this I'll be forced to input metadata for generating the CSV base files `config/bundle/bases/...` when I try to build and run my operator via `make manifest` or `make install`:

```console
$ operator-sdk init --repo=github.com/hasbro17/memcached-operator --domain=example.com
...

$ operator-sdk create api --group cache --version v1alpha1 --kind Memcached
...

$ make manifests
/Users/haseeb/work/go-space/bin//controller-gen "crd:trivialVersions=true" rbac:roleName=manager-role webhook paths="./..." output:crd:artifacts:config=config/crd/bases
operator-sdk generate bundle -q --kustomize

Display name for the operator (required):
...
```

This kind of forces CSVs and bundles into the workflow by default rather than making it opt-in like we had with `generate csv` or `generate bundle` in the old layout.

I wonder if we can't somehow decouple this from the core workflow of make manifests. Or make it optional so I don't get prompted for the interactive CSV metadata input when I don't want it.

Reply:
In my mind the question is: now that the SDK has integrated kubebuilder's scaffold/CLI, why use `operator-sdk` over `kubebuilder` if you don't want bundle (or package manifests) support?

Reply:
there's still the additional language/technology support

Reply:
Also related: https://github.com/operator-framework/enhancements/pull/16#issuecomment-634872330.

Reply:
@fabianvf agreed, but those are not integrated yet. I think it is reasonable to not scaffold bundle files in `init` for non-Go operators by default. However we need to have an easy path to add bundle support for those operator types. This EP should take those operator types into account _without_ discussing implementation just yet if possible.

Reply:
HI @estroz,  

> In my mind the question is: now that the SDK has integrated kubebuilder's scaffold/CLI, why use operator-sdk over kubebuilder if you don't want bundle (or package manifests) support?

IHMO:

- We do NOT need to change the default behaviour to allow the support for OLM or other SDK features 
- Also, allow users to integrate their projects with OLM does NOT necessarily mean be mandatory to do it
- I understand that the business strategy would still be keeping the scaffolded project working/compatible with Kubuilder which will NOT happen if we change the default makefile targets. I mean, to keep the backwards compatibility we would add things on top only but not change or delete what is scaffolded by default. 

The task of the PR operator-framework/operator-sdk#2860 is to add the make targets for the bundle, however, my concern is that we are NOT just adding new makefile targets there and we are changing the preexisting ones which are affecting the default expected behaviour and will breaking the backwards compatibility with kubebuilder.

Reply:
**Regards the operator-framework/operator-sdk#2860:**

It is to ADD the make targets for the bundle/OLM, however, my concern is that we are not just ADDING but we are UPDATING the default makefile targets there and affecting the expected behaviour and the backwards compatibility with kubebuilder besides force the user provides the CSV info when he/she should not care about it.

- **My concern is:** not about UX only. But we are breaking the backwards compatibility with kubebuilder which I understand that goes against the business strategy and making mandatory things that should not (optional). 


- **My suggestion is:** just let’s keep it simple and just ADD things on top instead update what is already scaffolded. We do not need to do it to achieve our goal.

I hope that I could clarify that.

Reply:
The argument for removing the `manifests` recipe amendment makes sense to me. Then if a user wants to generate a bundle and interactively fill in metadata, they would call `make bundle` or `make bundle-build`.

Just a note on wording:

>goes against the business strategy

There is no "business strategy" in the upstream SDK. We are promoting the operator-framework ecosystem which is why `operator-sdk` exists in the first place. We're not forcing operator developers to use bundles. We are saying "`operator-sdk` as an enhanced `kubebuilder` binary". If users want operator-framework features, then they will use `operator-sdk`. Having `operator-sdk` being a drop-in replacement for `kubebuilder` is fine too, I just don't see why you wouldn't just use `kubebuilder` if you're only using vanilla features.

Reply:
Hi @estroz, 

> The argument for removing the manifests recipe amendment makes sense to me. Then if a user wants to generate a bundle and interactively fill in metadata, they would call make bundle or make bundle-build.

It sounds great.  Please, feel free to ping me in the operator-framework/operator-sdk#2860 for a new review.

Reply:
I think I need some help understanding what the art of the possible is here. 

1. Are we able to make it "easy" to say 1 time that I want to be able to generate bundles and do that from then on?
2. Could that be a question/default during init?
3. If it was all in one, would someone coming from a kubebuilder built operator do this?
4. What if their operator was not a scaffold that we expect? Could we still set them up to generate this info?

Reply:
Going to answer these out of order for better flow:

3,4: The best course of action to upgrade from a vanilla kubebuilder project would be documenting how to add bundle support. Furthermore no `operator-sdk` commands should assume bundle-related files exist; then `operator-sdk` can still be used as a drop-in replacement for `kubebuilder`-scaffolded projects.

1,2: If we ask the user during `operator-sdk init`, they say no, then want to add support later, they'd have to follow the same documentation for as kubebuilder projects, which isn't smooth. Instead what I inferred from @camilamacedo86's point is we should always scaffold all modifications to the Makefile on running `operator-sdk init` so that as soon as a user runs `make bundle` or `make bundle-build` all bundle files are available, and not beforehand.

The main change to this EP if we're OK with the above is to remove the additional `generate bundle` line from the `manifests` make recipe.